includes/Core/Util/Migration_1_128_0.php

```diff
@@ -0,0 +1,187 @@
+use Google\Site_Kit\Modules\Ads\Settings as Ads_Settings;
+
+/**
+ * Class Migration_1_128_0
+ *
+ * @since 1.128.0
+ * @access private
+ * @ignore
+ */
+class Migration_1_128_0 {
+	/**
+	 * Target DB version.
+	 */
+	const DB_VERSION = '1.128.0';
+
+	/**
+	 * DB version option name.
+	 */
+	const DB_VERSION_OPTION = 'googlesitekit_db_version';
+
+	/**
+	 * Analytics options name.
+	 */
+	const ANALYTICS_OPTIONS = 'googlesitekit_analytics-4_settings';
+
+	/**
+	 * Ads options name.
+	 */
+	const ADS_OPTIONS = 'googlesitekit_ads_settings';
+
+	/**
+	 * Context instance.
+	 *
+	 * @since 1.128.0
+	 * @var Context
+	 */
+	protected $context;
+
+	/**
+	 * Options instance.
+	 *
+	 * @since 1.128.0
+	 * @var Options
+	 */
+	protected $options;
+
+	/**
+	 * Analytics_Settings instance.
+	 *
+	 * @since 1.128.0
+	 * @var Analytics_Settings
+	 */
+	protected $analytics_settings;
+
+	/**
+	 * Ads_Settings instance.
+	 *
+	 * @since 1.128.0
+	 * @var Ads_Settings
+	 */
+	protected $ads_settings;
+
+	/**
+	 * Constructor.
+	 *
+	 * @since 1.128.0
+	 *
+	 * @param Context $context Plugin context instance.
+	 * @param Options $options Optional. Options instance.
+	 */
+	public function __construct(
+		Context $context,
+		Options $options = null
+	) {
+		$this->context            = $context;
+		$this->options            = $options ?: new Options( $context );
+		$this->analytics_settings = new Analytics_Settings( $this->options );
+		$this->ads_settings       = new Ads_Settings( $this->options );
+	}
+
+	/**
+	 * Registers hooks.
+	 *
+	 * @since 1.128.0
+	 */
+	public function register() {
+		add_action( 'admin_init', array( $this, 'migrate' ) );
+	}
+
+	/**
+	 * Migrates the DB.
+	 *
+	 * @since 1.128.0
+	 */
+	public function migrate() {
+		$db_version = $this->options->get( self::DB_VERSION_OPTION );
+
+		if ( ! $db_version || version_compare( $db_version, self::DB_VERSION, '<' ) ) {
+			$this->migrate_analytics_conversion_id_setting();
+			$this->activate_ads_module();
+
+			$this->options->set( self::DB_VERSION_OPTION, self::DB_VERSION );
+		}
+	}
+
+	/**
+	 * Migrates the Ads Conversion ID to the new Ads module.
+	 *
+	 * @since 1.128.0
+	 */
+	protected function migrate_analytics_conversion_id_setting() {
+		if ( ! $this->analytics_settings->has() ) {
+			return;
+		}
+
+		$analytics_settings = $this->options->get( self::ANALYTICS_OPTIONS );
+
+		if ( empty( $analytics_settings ) || ! array_key_exists( 'adsConversionID', $analytics_settings ) || empty( $analytics_settings['adsConversionID'] ) ) {
+			return;
+		}
+
+		$ads_settings = $this->options->get( self::ADS_OPTIONS );
+
+		if ( array_key_exists( 'conversionID', $ads_settings ) && ! empty( $ads_settings['conversionID'] ) ) {
+			// If there is already an adsConversionID set in the Ads module, do not overwrite it, remove it from the Analytics module.
+			$this->options->set( self::ANALYTICS_OPTIONS, array_merge( $analytics_settings, array( 'adsConversionID' => '' ) ) );
```

Comment:
If we want to delete a property, we should use `unset`:

```php
unset( $analytics_settings['adsConversionID'] );
$this->options->set( self::ANALYTICS_OPTIONS, $analytics_settings );
```

Reply:
Plus, we need to remove the `adsConversionIDMigratedAtMs` option as well.

Reply:
Why do we need to remove `adsConversionIDMigratedAtMs`? It should never have been set in this case, where the this migration is running and there is already a value in the Ads module, unless the frontend migration has been run by the user, in which case they should still see the banner using  `adsConversionIDMigratedAtMs` based on the existing frontend dismiss logic.

Reply:
Ok, sounds good.